### PR TITLE
File Uploads: Image endpoint

### PR DESCRIPTION
**Pull request by johnnyomair**
---

Open questions:

- [x] How to share code between DAM and File Uploads -> Reduced duplication in cb1f97f4496b71dd89d17a57f781595e09e740f3, will create an ImgproxyModule in the future. Task: https://vivid-planet.atlassian.net/browse/COM-1107

## PR Checklist

-   [x] Verify if the change requires a changeset. See [CONTRIBUTING.md](https://github.com/vivid-planet/comet/blob/HEAD/CONTRIBUTING.md)
     Added to original changeset
-   [x] Link to the respective task if one exists: COM-1002
-   [x] Provide screenshots/screencasts if the change contains visual changes

<details>
    <summary>Screenshots/screencasts</summary>
    
file uploads image preview

</details>

### CODE REVIEW

**Comment by johnnyomair** [comment]
I've copied code from the DAM module to make this work. Also, I use the imgproxy and scaled images module from the DAM. Should we create a third module that's used by both the DAM and the file uploads module?

File: demo/admin/src/products/ProductForm.tsx [hunk 1]
```diff
@@ -331,6 +331,8 @@ export function ProductForm({ id }: FormProps) {
                             multiple
                             maxFileSize={1024 * 1024 * 4} // 4 MB
                             fullWidth
+                            layout="grid"
+                            previewImageWidth={640}
```
Comment (jamesricky) [inline demo/admin/src/products/ProductForm.tsx:335]: I don't think setting this value should be necessary, maybe it shouldn't even be an option. 
Each Item as a min-width of 200px, so a little more (~250) multiplied by 2 (hi-dpi) should probably be set as the default for this.

https://github.com/vivid-planet/comet/blob/46f93229999393a28673718cb289a3b69707ba15/packages/admin/admin/src/form/file/FileSelect.tsx#L214

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:335]: It isn't necessary, I'm currently using 320 as default (should this be 640?). However, as allowed image resize widths can be defined in the application, it needs to be possible to override the default value.

Comment (jamesricky) [reply demo/admin/src/products/ProductForm.tsx:335]: Ah ok 👍🏼 

I think the maximum size the items can have is 230px, so the api should deliver at least twice that width. 
So 460, or whatever the next largest value is, that the api supports. Probably 640 then 👍🏼

File: packages/admin/cms-admin/src/form/file/FinalFormFileUpload.tsx [hunk 2]
```diff
@@ -46,11 +48,14 @@ type FinalFormFileUploadMultipleFilesProps = FieldRenderProps<GQLFinalFormFileUp
...
     : FinalFormFileUploadSingleFileProps) &
-    Partial<FileSelectProps<GQLFinalFormFileUploadFragment>>;
+    Partial<FileSelectProps<GQLFinalFormFileUploadFragment>> & {
+        previewImageWidth?: number;
+    };
 
 export const FinalFormFileUpload = <Multiple extends boolean | undefined>({
     input: { onChange, value: fieldValue, multiple },
     maxFiles,
+    previewImageWidth = 320,
```
Comment (johnnyomair) [inline packages/admin/cms-admin/src/form/file/FinalFormFileUpload.tsx:58]:
```suggestion
    previewImageWidth = 640,
```

File: demo/admin/src/products/ProductForm.tsx (hunk 1 above)
Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:335]: Changed: 1015071ad52af29d082e97fb1e90158d76a36206

**Review by jamesricky — APPROVED** [review]

File: demo/api/schema.gql [hunk 3]
```diff
@@ -186,6 +186,7 @@ type FileUpload {
   createdAt: DateTime!
   updatedAt: DateTime!
   downloadUrl: String
+  previewUrlTemplate: String
```
Comment (nsams) [inline demo/api/schema.gql:189]: Is this really only a "preview"? Yes, in this PR you use it as preview, but isn't that also to be used for example to show scaled User Profile Images?

And the template thing:

We have that in dam as we use those images as responsive images with a large number of different sizes, and we don't want to include all individual urls in the api response - so we create them client side using the template

Is that also the use case for FileUpload images? (alternative could be a width argument)

Comment (johnnyomair) [reply demo/api/schema.gql:189]: > (alternative could be a width argument)

Isn't that what I'm doing? Or do you mean a Get-Parameter?

> Is this really only a "preview"? Yes, in this PR you use it as preview, but isn't that also to be used for example to show scaled User Profile Images?

You're right, there are more use cases. We could call it `imageUrl` instead.

File: demo/admin/src/products/ProductForm.tsx [hunk 4]
```diff
@@ -331,6 +331,7 @@ export function ProductForm({ id }: FormProps) {
                             multiple
                             maxFileSize={1024 * 1024 * 4} // 4 MB
                             fullWidth
+                            layout="grid"
```
Comment (thomasdax98) [inline demo/admin/src/products/ProductForm.tsx:334]: I'm not sure about `layout=grid` being the switch between preview or preview-less mode. I'd prefer something more explicit like `previewImages=true`

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:334]: @jamesricky what do you think?

File: demo/api/schema.gql (hunk 3 above)
Comment (johnnyomair) [reply demo/api/schema.gql:189]: However, if we want to use it for scaled user profile images, they should probably be without a timeout.

**Comment by johnnyomair** [comment]
@nsams @thomasdax98 incorporated the changes we discussed yesterday. Now the only open concern is https://github.com/vivid-planet/comet/pull/2417#discussion_r1770901920.

File: demo/admin/src/products/ProductForm.tsx (hunk 4 above)
Comment (jamesricky) [reply demo/admin/src/products/ProductForm.tsx:334]: Do you mean we might need to show an actual preview-image in the list-view in the future? 
I don't think that will be the case. At least all designs only show the preview in the grid view.

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:334]: > Do you mean we might need to show an actual preview-image in the list-view in the future?

Yes.

>  I don't think that will be the case. At least all designs only show the preview in the grid view.

It's not the first time that designs have changed considerably 😁 Maybe the next layout will be "imageCarousel" or whatever. We'd than need an additional showAsCarouselWithPreview

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:334]: Alternative idea: Switch to layout="grid" automatically if any of the files has a preview URL and layout hasn't been explicitly set to "list"?

Comment (jamesricky) [reply demo/admin/src/products/ProductForm.tsx:334]: > Alternative idea: Switch to layout="grid" automatically if any of the files has a preview URL and layout hasn't been explicitly set to "list"?

The designers created this as two separate components actually: 
- A file upload, which includes a list of files that have been uploaded
- A grid of preview images

They never intended combining those in any way, so I think it should explicitly be set which variant to show. 

The reason we combined them (initially I think they were separate) was that they both needed to access the data in the form state and handle that data in the same way. 

Maybe it makes sense to separate them after all 🤔 

> Maybe the next layout will be "imageCarousel"

That is actually in the design and will probably come in the future but it's part of the grid variant, and opens when you click on one of the images in a sort of modal 😄

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:334]: So showAsGridWithPreview wouldn't solve the problem either? 😬 

Maybe we should change this in a follow-up PR?

Comment (jamesricky) [reply demo/admin/src/products/ProductForm.tsx:334]: > Maybe we should change this in a follow-up PR?

Yes, maybe just leave it as is for now and we can discuss this in person another time

Comment (johnnyomair) [reply demo/admin/src/products/ProductForm.tsx:334]: @thomasdax98 would you be okay with this?

**Review by jamesricky — APPROVED** [review]

File: demo/admin/src/products/ProductForm.tsx (hunk 4 above)
Comment (thomasdax98) [reply demo/admin/src/products/ProductForm.tsx:334]: That's fine by me

**Review by thomasdax98 — APPROVED** [review]

**Pull request merged**